src/luaotfload-harf-define.lua

```diff
@@ -0,0 +1,372 @@
+      local extents = hbfont:get_glyph_extents(gid)
+      if extents then
+        height = extents.y_bearing
+        depth = extents.y_bearing + extents.height
+        if extents.x_bearing < 0 then
+          italic = -extents.x_bearing
+        end
+      end
+      glyphs[gid] = {
+        width  = width,
+        height = height or ascender,
+        depth  = -(depth or descender),
+        italic = italic or 0,
+      }
+    end
+
+    local unicodes = hbface:collect_unicodes()
+    local characters = {}
+    local nominals = {}
+    for _, uni in next, unicodes do
+      local glyph = hbfont:get_nominal_glyph(uni)
+      characters[uni] = glyph
+      nominals[glyph] = uni
+    end
+
+    local xheight, capheight = 0, 0
+    if hasos2 then
+      local os2 = hbface:get_table(os2tag)
+      local length = os2:get_length()
+      local data = os2:get_data()
+      if length >= 96 and string.unpack(">H", data) > 1 then
+        -- We don’t need much of the table, so we read from hard-coded offsets.
+        xheight = string.unpack(">H", data, 87)
+        capheight = string.unpack(">H", data, 89)
+      end
+    end
+
+    if xheight == 0 then
+      local gid = characters[120] -- x
+      if gid then
+        xheight = glyphs[gid].height
+      else
+        xheight = ascender / 2
+      end
+    end
+
+    if capheight == 0 then
+      local gid = characters[88] -- X
+      if gid then
+        capheight = glyphs[gid].height
+      else
+        capheight = ascender
+      end
+    end
+
+    data = {
+      gid_offset = 0x120000,
+      face = hbface,
+      font = hbfont,
+      upem = upem,
+      fonttype = fonttype,
+      space = space,
+      xheight = xheight,
+      capheight = capheight,
+      slant = slant,
+      glyphs = glyphs,
+      nominals = nominals,
+      unicodes = characters,
+      psname = hbface:get_name(hb.ot.NAME_ID_POSTSCRIPT_NAME),
+      fullname = hbface:get_name(hb.ot.NAME_ID_FULL_NAME),
+      haspng = hbface:ot_color_has_png(),
+      loaded = {}, -- Cached loaded glyph data.
+    }
+
+    hbfonts[key] = data
+    return data
+  end
+end
+
+-- Drop illegal characters from PS Name, per the spec
+-- https://docs.microsoft.com/en-us/typography/opentype/spec/name#nid6
+local function sanitize(psname)
+  return psname:gsub('[][\0-\32\127-\255(){}<>/%%]', '-')
+end
+
+-- Ligatures. The value is a character "ligature" table as described in the
+-- manual.
+local tlig ={
+  [0x2013] = { [0x002D] = { char = 0x2014 } }, -- [---]
+  [0x002D] = { [0x002D] = { char = 0x2013 } }, -- [--]
+  [0x0060] = { [0x0060] = { char = 0x201C } }, -- [``]
+  [0x0027] = { [0x0027] = { char = 0x201D } }, -- ['']
+  [0x0021] = { [0x0060] = { char = 0x00A1 } }, -- [!`]
+  [0x003F] = { [0x0060] = { char = 0x00BF } }, -- [?`]
+  [0x002C] = { [0x002C] = { char = 0x201E } }, -- [,,]
+  [0x003C] = { [0x003C] = { char = 0x00AB } }, -- [<<]
+  [0x003E] = { [0x003E] = { char = 0x00BB } }, -- [>>]
+}
```

Comment:
I used global variables for `tlig` and `trep` with the intention that people would use them to implement ligatures in a similar (but limited) way to what the ConTeXt based loader provides. I’m not too attached to this, but I recall many queries from people wanting to patch fonts on the fly and add missing ligature records. I think it would be nice to keep this functionality in some way or another.

Reply:
This can be implemented using the fontloaders mechanism for user defined features. These tables became local especially to avoid modifications: User-defined ligatures should in general not be controlled by `tlig` but by a custom feature, otherwise they will often affect more fonts than intended.

Reply:
I agree, but there is no other mechanism for HarfBuzz-processed fonts, any code that modifies the Lua tables that are used by the other loader will not work here. Another table/option/feature is needed, it does not have to be `tlig`.

Reply:
The fontloader provides multiple kinds of user-defined features, there are the higher-level ones which reuse the fontloaders OpenType feature handling and there is are features on a lower level which can modify the font table or modify each nodelist which contains the font. The last mechanism is e.g. used in luaotfload-colors.lua to implement the `color` feature. These low-level feature work with HarfBuzz fonts too.

These could be used to implement something like `tlig` by modifying the font table to add ligatures exactly like it is done here and `trep` could be implemented by pre-processing the node list and rewriting the affected nodes.

We could add a nicer interface for this.